anvill/src/Analysis/CrossReferenceResolver.cpp

```diff
@@ -698,9 +698,18 @@ void CrossReferenceResolver::ClearCache(void) const {
   impl->xref_cache.clear();
 }
 
-// Try to resolve `val` as a cross-reference.
+// Try to resolve `val` as a cross-reference. `uses_cache` flag is set to true
+// if the application is using the cache and does not want to invalidate it.
 ResolvedCrossReference
-CrossReferenceResolver::TryResolveReference(llvm::Value *val) const {
+CrossReferenceResolver::TryResolveReference(llvm::Value *val,
+                                            bool uses_cache) const {
+
+  // If the application is not using cache, invalidate it before resolving
+  // the cross references. It is done to avoid stale `val` sitting in the
+  // cache if it has been changed/deleted.
+  if (!uses_cache) {
```

Comment:
I don't like this `bool` parameter; I think perhaps what you could do is instead rename the method, and add a method:

```
TryResolveReferenceWithCaching
TryResolveReferenceWithClearedCache
```

Or something like this. That, or pass in an explicit cache datastructure, and then maybe `xref_cache` a pointer to it, that you can set/reset. E.g.:

```
using CrossReferenceCache = std::unordered_map<llvm::Value *, ResolvedCrossReference>;
```

```
CrossReferenceCache *xref_cache{nullptr};

...

ResolvedCrossReference
CrossReferenceResolver::TryResolveReference(llvm::Value *val,
                                            CrossReferenceCache &xref_cache_) const {
  impl->xref_cache = &xref_cache_;
  auto xr = impl->ResolveValue(val);
  impl->xref_cache = nullpltr;
  return xr;
}
```

Then above you do things like:
```
ResolvedCrossReference dummy_xr;
ResolvedCrossReference *cached_xr = &dummy_xr;
if (xref_cache) {
  auto it = xref_cache->find(val);
  if (it != xref_cache->end()) {
    return it->second;
  }
  cached_xr = &(xref_cache->emplace(val, dummy_xr).first->second);
}
auto &xr = *cached_xr;
```

Reply:
I updated the function name to:
```
      TryResolveReferenceWithCaching
       TryResolveReferenceWithClearedCache
```
It will be nice for `CrossReferenceResolver` to always use cache internally if it save analysis time. If the application wants to access the cache, we can add an interface function to expose it.